### PR TITLE
Testing the code in the paper

### PR DESCRIPTION
**Pull request by rsln-s**
Adding the tests for snippets in the upcoming paper "Fast Simulation of High-Depth QAOA Circuits" describing QOKit.

### CODE REVIEW

**Comment by rsln-s** [comment]
`test_listing_3` is perhaps OK to skip, but `test_listing_2` must pass. Currently, I get the following error on a GPU instance:

```
======================================================== FAILURES ========================================================
_____________________________________________________ test_listing_2 _____________________________________________________

    def test_listing_2():
        import qokit
    
        simclass = qokit.fur.choose_simulator_xycomplete()
        n = 8
        terms = qokit.labs.get_terms(n)
>       sim = simclass(n, terms=terms)

tests/test_examples_in_the_paper.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
qokit/fur/nbcuda/qaoa_simulator.py:24: in __init__
    super().__init__(n_qubits, costs, terms)
qokit/fur/qaoa_simulator_base.py:63: in __init__
    self._hc_diag = self._diag_from_terms(terms)
qokit/fur/nbcuda/qaoa_simulator.py:32: in _diag_from_terms
    precompute_gpu(0, self.n_qubits, terms, out)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

rank = 0, n_local_qubits = 8, terms = [(0, 1, 3, 4), (0, 1, 4, 5), (0, 2, 3, 5), (3, 7), (4, 6), (5, 7), ...]
out = <numba.cuda.cudadrv.devicearray.DeviceNDArray object at 0x7f138519b7c0>

    def compute_costs(rank: int, n_local_qubits: int, terms, out):
        offset = rank << n_local_qubits
        n = len(out)
        zero_init_kernel.forall(n)(out)
    
>       for coef, pos in terms:
E       ValueError: too many values to unpack (expected 2)

qokit/fur/diagonal_precomputation/gpu_numba.py:39: ValueError
```

**Comment by rsln-s** [comment]
The same issue happens in CPU-only as evidenced by the GitHub workflow logs:

```
...
>       for coeff, pos in weighted_terms:
E       ValueError: too many values to unpack (expected 2)
...
```

File: qokit/labs.py [hunk 1]
```diff
@@ -579,6 +601,4 @@ def get_gate_optimized_terms_greedy(N: int, number_of_gate_zones: int = 4, seed:
                 j += 1
             k += swapped
 
-    assert set(circuit) == set(get_terms(N))
```
Comment (rsln-s) [inline qokit/labs.py:582]: Any reason to drop this check? Now that `get_terms` returns weights well, should this become `assert set(circuit) == set(get_term_indices(N))` or something like that?

**Review by rsln-s — COMMENTED** [review]
LGTM modulo the removal of assertion in `qokit/labs.py`

File: qokit/labs.py (hunk 1 above)
Comment (danlkv) [reply qokit/labs.py:582]: This is already checked with [tests](https://github.com/jpmorganchase/QOKit/blob/46b1f517a3c7014989dd5f724c056b9ff79877f9/tests/test_labs.py#L51-L52), so runinng `get_terms` seemed redundant. We can keep the assert but then the tests are a bit silly:)

Comment (rsln-s) [reply qokit/labs.py:582]: Ah, right! Thanks for catching

**Pull request merged**